README
* What is VTE?

VTE is a library (libvte) implementing a terminal emulator widget for GTK+,
and a minimal sample application (vte) using that.  Vte is mainly used in
gnome-terminal, but can also be used to embed a console/terminal in games,
editors, IDEs, etc.

Available versions:
  
  https://developer.gnome.org/vte/

VTE doesn't have a homepage.  Report any issues at:

  http://bugzilla.gnome.org/enter_bug.cgi?product=vte


NEWS
0.45.90
=======

* VTE now depends on PCRE2, and GRegex (which uses PCRE1 internally)
use is deprecated.

0.37.0
======
* New API: vte-2.91
* Many improvements and bug fixes
* Translation updates

0.36.0
======
* Build fixes (#726604)
* Docs updates

0.35.90
=======
* prompt: Use precmd hook on zsh (#724982)
* Fix many compiler warnings
* Remove obsolete gtk+-2.0 conditional code
* Support OSC 19 and 119 sequences (#725974)
* Also pass Ctrl-F1 to applications (#726438)
* Nicer box drawing characters (#709556)

0.35.2
======
* Colour palette handling fixes (#640040)
* Add support for OSC 104 (reset colors) (#640040)
* Add support for OSC 1?1[017] (fg, bg, highlight colors) (#640040)
* Add support for OSC 112 (reset cursor color) (#640040)
* Make OSC response use the same terminator as the query (#722446)
* Maintain selection when rewrapping (#722635)
* Fix selection with widget padding adjustment (#722873)

0.35.1
======
* Support for 16 million colors (#704449)
* Fix Home/End and some other keys (#600659, #643233, #710349)
* Add API to enable/disable rewrap on resize (#336238)
* Fix a longstanding performance regression (#721944)
* Bug fixes (#586950, #619891, #672210, #685759, #691972, #720977, #721522)

0.35.0
======
* Build fixes
* Portability fixes (#710165)
* Rewrap contents upon window resize (#336238)
* Bug fixes (#4993, #13299, #542087, #686097, #693013, #705985, #709692,
  #709584, #709930, #710426, #710481, #710483, #710484, #710488, #710486)

Many thanks to Egmont Koblinger who contributed most of these bug fixes!

0.34.9
======
* Lines disappear after positioning back the cursor (#708496)
* Garbage displayed after resizing whilst using an alternate screen (#415277)
* vte uses too many file descriptors (part of #646098)
* Incorrect scrollbar position after the "less" command (#676075)
* Add support for DEC 1007 to set the alternate scroll mode (#518405, #709060)
* Use O_TMPFILE when available (#704705)
* Improvements to scrollback storage (#637275) 
* Aborts on assertion (#676075)
* zsh - lots of blank space upon resizing VTE based terminals (#708213)
* Misc fixes

0.34.8
======
- Build fixes
- Fix crash on pathological fonts (#708196)
- Use correct character attributes around graphic characters (part of #707221)
- Make extended fore/background colour handling more xterm-like (#616436)

0.34.7
======
- Honour gtk-enable-primary-paste setting
- Redraw when changing colours (#702415)

0.34.6
======
- vte.sh bugfix
- Add new flag for vte_terminal_fork_command_full() to not copy the 
  parent  environment
- Allow the empty string to OSC 6 and OSC 7 to unset a previously
  set value

0.34.5
======
- Improved vte.sh script (#697475): It is now not necessary anymore
  to manually change PS1. On distributions that automatically source
  /etc/profile.d/*.sh scripts for all interactive shells (ie Fedora),
  no manual intervention is needed at all; for others the only change
  required is to source the vte.sh script.
  Also, the script now works for both bash and zsh shells.

0.34.4
======
- Improved vte.sh script (#697225)

0.34.3
======
- Introspection fixes (#676999, #690003)
- Build fixes (#690554)
- Portability fixes (#694494)
- Translation updates

0.34.2
======
- Translation updates

0.34.1
======
* End selection mode misses Shift release (#683730)
* Don't report shift-left-button release (#683729)
* xterm extended mouse tracking mode (#681329)
* Support SGR 3 (italic text) (#685223)

0.34.0
======
* Support escape sequences to set the current directory (#675987)
* Fixed drawing of Box Drawing and Block Elements characters (#682692)
* Introspection fixes (#679805, #681714)
* Build fixes

0.32.2
======
- emulation: Fix DoS by malicious escape sequences (#676090)
- widget: Fix crash with NULL tag in match_check (#676886)
- keymap: Treat ALT as META (#663779)
 
0.32.1
======
- pty: Fix custom child setup function (#672649)
- leak fix in vte_draw_set_text_font() (#672757)
- emulation: Fix off-by-one error in interpretation of ECH (#670037)

0.32.0
======
- Use GResource for our default termcap
- build fixes for interix (#652290)
- Allocate PTYs with openpty on BSD (#670758)
- Introspection fixes (#648183, #655402, #661716)
- Fix mouse whell scrolling with gtk+ 3.4 (#671305)
- Use correct library name for glade integration (#671760)
- gnome-pty-helper: Respect SUID_CFLAGS and SUID_LDFLAGS (#670732)

0.31.0
======
- Don't override the widget background colour
- Add missing (closure) annotation on VteSelectionFunc (#655405, Evan Nemerson)
- use gtk_drag_check_threshold() (#654691, Dan Winship)
- Add urxvt extended mouse tracking mode support (#662423, Egmont Koblinger)
- Add runtime check for X11 display (#660151, Young-Ho Cha)
- Build fixes (#661121, #616001)
- Translation updates

0.30.1
======
- Translation updates

0.30.0
======
- Translation updates

0.29.1
======
- Bugs fixed:
  - Bug 657581 - gdk-3.0 supports multiple targets; the "target" variable is gone
  - Bug 657584 - vte needs to properly set VTE_API_VERSION

0.29.0
======
- Drop GTK 2 support
- Fix accessibility with GTK 3.1 (#654630)
- Translation updates

0.28.1
======
- Bugs fixed:
   * Bug 652124 - malicious escape sequence causes gnome-terminal to exhaust memory
- Translation updates

0.28.0
======
- Bugs fixed:
   * Bug 642184 - vte does not build against new glade (with gladeui-2.0)
- Translation updates

0.27.90
=======
- Fix shrinking terminal on gtk3
- Translation updates

0.27.5
======
- Fix build with recent GTK+
- Translation updates

0.27.4
======
- Fix build with recent GTK+
- Introspection fixes for fork_command_full
- Translation updates

0.27.3
======
- Add a gsettings .enums.xml file for vte's enum types
- Fix build with latest gtk 3
- Fix introspection annotations and require gobject-introspection 0.9.0
- Bugs fixed:
  * Bug 628870 - Clarify error message
- Translation updates

0.27.2
======
(See git log)

0.27.1
======

- Bugs fixed:

src/caps.c
/*
 * Copyright (C) 2001,2002 Red Hat, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <config.h>
#include <stdlib.h>
#include <glib.h>
#include "caps.h"

#define ESC _VTE_CAP_ESC
#define CSI _VTE_CAP_CSI
#define ST  _VTE_CAP_ST
#define OSC _VTE_CAP_OSC
#define PM  _VTE_CAP_PM
#define APC _VTE_CAP_APC

#define ENQ "\005"
#define BEL "\007"
#define BS  "\010"
#define TAB "\011"
#define LF  "\012"
#define VT  "\013"
#define FF  "\014"
#define CR  "\015"
#define SO  "\016"
#define SI  "\017"
#define DEL "\177"

#define ENTRY(seq, name) seq "\0" name "\0"
#define COMMENT(c)

/* From some really old XTerm docs we had at the office, and an updated
 * version at Moy, Gildea, and Dickey. */
const char _vte_xterm_capability_strings[] =
        ENTRY(ENQ, "return-terminal-status")
        ENTRY(BEL, "bell")
        ENTRY(BS,  "backspace")
        ENTRY(TAB, "tab")
        ENTRY(LF,  "line-feed")
        ENTRY(VT,  "vertical-tab")
        ENTRY(FF,  "form-feed")
        ENTRY(CR,  "carriage-return")
        ENTRY(SO,  "shift-out")
        ENTRY(SI,  "shift-in")
        ENTRY(DEL, "backspace")

        ENTRY(ESC " F", "7-bit-controls")
        ENTRY(ESC " G", "8-bit-controls")
        ENTRY(ESC " L", "ansi-conformance-level-1")
        ENTRY(ESC " M", "ansi-conformance-level-2")
        ENTRY(ESC " N", "ansi-conformance-level-3")
        ENTRY(ESC "#3", "double-height-top-half")
        ENTRY(ESC "#4", "double-height-bottom-half")
        ENTRY(ESC "#5", "single-width")
        ENTRY(ESC "#6", "double-width")
        ENTRY(ESC "#8", "screen-alignment-test")

        COMMENT(/* These are actually designate-other-coding-system from ECMA 35,)
                   COMMENT( * but we don't support the full repertoire.  Actually, we don't)
                   COMMENT( * know what the full repertoire looks like. */)
        ENTRY(ESC "%%@", "default-character-set")
        ENTRY(ESC "%%G", "utf-8-character-set")

        ENTRY(ESC "(0", "designate-g0-line-drawing")
        ENTRY(ESC "(A", "designate-g0-british")
        ENTRY(ESC "(B", "designate-g0-plain")
        ENTRY(ESC ")0", "designate-g1-line-drawing")
        ENTRY(ESC ")A", "designate-g1-british")
        ENTRY(ESC ")B", "designate-g1-plain")

        ENTRY(ESC "7", "save-cursor")
        ENTRY(ESC "8", "restore-cursor")
        ENTRY(ESC "=", "application-keypad")
        ENTRY(ESC ">", "normal-keypad")
        ENTRY(ESC "D", "index")
        ENTRY(ESC "E", "next-line")
        COMMENT(/* ENTRY(ESC "F", "cursor-lower-left") */)
        ENTRY(ESC "H", "tab-set")
        ENTRY(ESC "M", "reverse-index")
        COMMENT(/* ENTRY(ESC "N", "single-shift-g2") */)
        COMMENT(/* ENTRY(ESC "O", "single-shift-g3") */)
        ENTRY(ESC "P%s" ST, "device-control-string")
        ENTRY(ESC "V", "start-of-guarded-area")
        ENTRY(ESC "W", "end-of-guarded-area")
        ENTRY(ESC "X%s" ST, "start-of-string / end-of-string")
        ENTRY(ESC "Z", "return-terminal-id")
        ENTRY(ESC "c", "full-reset")
        ENTRY(ESC "l", "memory-lock")
        ENTRY(ESC "m", "memory-unlock")
        COMMENT(/* ENTRY(ESC "n", "invoke-g2-character-set") */)
        COMMENT(/* ENTRY(ESC "o", "invoke-g3-character-set") */)
        COMMENT(/* ENTRY(ESC "|", "invoke-g3-character-set-as-gr") */)
        COMMENT(/* ENTRY(ESC "}", "invoke-g2-character-set-as-gr") */)
        COMMENT(/* ENTRY(ESC "~", "invoke-g1-character-set-as-gr") */)

        COMMENT(/* APC stuff omitted. */)

        COMMENT(/* DCS stuff omitted. */)

        ENTRY(CSI "@", "insert-blank-characters")
        ENTRY(CSI "%d@", "insert-blank-characters")
        ENTRY(CSI "A", "cursor-up")
        ENTRY(CSI "%dA", "cursor-up")
        ENTRY(CSI "B", "cursor-down")
        ENTRY(CSI "%dB", "cursor-down")
        ENTRY(CSI "C", "cursor-forward")
        ENTRY(CSI "%dC", "cursor-forward")
        ENTRY(CSI "D", "cursor-backward")
        ENTRY(CSI "%dD", "cursor-backward")
        ENTRY(CSI "E", "cursor-next-line")
        ENTRY(CSI "%dE", "cursor-next-line")
        ENTRY(CSI "F", "cursor-preceding-line")
        ENTRY(CSI "%dF", "cursor-preceding-line")
        ENTRY(CSI "G", "cursor-character-absolute")
        ENTRY(CSI "%dG", "cursor-character-absolute")
        ENTRY(CSI "H", "cursor-position")
        ENTRY(CSI ";H", "cursor-position")
        ENTRY(CSI "%dH", "cursor-position")
        ENTRY(CSI "%d;H", "cursor-position")
        ENTRY(CSI ";%dH", "cursor-position-top-row")
        ENTRY(CSI "%d;%dH", "cursor-position")
        ENTRY(CSI "I", "cursor-forward-tabulation")
        ENTRY(CSI "%dI", "cursor-forward-tabulation")
        ENTRY(CSI "J", "erase-in-display")
        ENTRY(CSI "%mJ", "erase-in-display")
        ENTRY(CSI "?J", "selective-erase-in-display")
        ENTRY(CSI "?%mJ", "selective-erase-in-display")
        ENTRY(CSI "K", "erase-in-line")
        ENTRY(CSI "%mK", "erase-in-line")
        ENTRY(CSI "?K", "selective-erase-in-line")
        ENTRY(CSI "?%mK", "selective-erase-in-line")
        ENTRY(CSI "L", "insert-lines")
        ENTRY(CSI "%dL", "insert-lines")
        ENTRY(CSI "M", "delete-lines")
        ENTRY(CSI "%dM", "delete-lines")
        ENTRY(CSI "P", "delete-characters")
        ENTRY(CSI "%dP", "delete-characters")
        ENTRY(CSI "S", "scroll-up")
        ENTRY(CSI "%dS", "scroll-up")
        ENTRY(CSI "T", "scroll-down")
        ENTRY(CSI "%dT", "scroll-down")
        ENTRY(CSI "%d;%d;%d;%d;%dT", "initiate-hilite-mouse-tracking")
        ENTRY(CSI "X", "erase-characters")
        ENTRY(CSI "%dX", "erase-characters")
        ENTRY(CSI "Z", "cursor-back-tab")
        ENTRY(CSI "%dZ", "cursor-back-tab")

        ENTRY(CSI "`", "character-position-absolute")
        ENTRY(CSI "%d`", "character-position-absolute")
        ENTRY(CSI "b", "repeat")
        ENTRY(CSI "%db", "repeat")
        ENTRY(CSI "c", "send-primary-device-attributes")
        ENTRY(CSI "%dc", "send-primary-device-attributes")
        ENTRY(CSI ">c", "send-secondary-device-attributes")
        ENTRY(CSI ">%dc", "send-secondary-device-attributes")
        ENTRY(CSI "=c", "send-tertiary-device-attributes")
        ENTRY(CSI "=%dc", "send-tertiary-device-attributes")
        ENTRY(CSI "?%mc", "linux-console-cursor-attributes")
        ENTRY(CSI "d", "line-position-absolute")
        ENTRY(CSI "%dd", "line-position-absolute")
        ENTRY(CSI "f", "cursor-position")
        ENTRY(CSI ";f", "cursor-position")
        ENTRY(CSI "%df", "cursor-position")
        ENTRY(CSI "%d;f", "cursor-position")
        ENTRY(CSI ";%df", "cursor-position-top-row")
        ENTRY(CSI "%d;%df", "cursor-position")
        ENTRY(CSI "g", "tab-clear")
        ENTRY(CSI "%dg", "tab-clear")

        ENTRY(CSI "%mh", "set-mode")
        ENTRY(CSI "?%mh", "decset")

        ENTRY(CSI "%mi", "media-copy")
        ENTRY(CSI "?%mi", "dec-media-copy")

        ENTRY(CSI "%ml", "reset-mode")
        ENTRY(CSI "?%ml", "decreset")

        ENTRY(CSI "%mm", "character-attributes")

        ENTRY(CSI "%dn", "device-status-report")
        ENTRY(CSI "?%dn", "dec-device-status-report")
        ENTRY(CSI "!p", "soft-reset")
        ENTRY(CSI "%d;%d\"p", "set-conformance-level")
        ENTRY(CSI " q", "set-cursor-style")
        ENTRY(CSI "%d q", "set-cursor-style")
        ENTRY(CSI "%d\"q", "select-character-protection")

        ENTRY(CSI "r", "set-scrolling-region")
        ENTRY(CSI ";r", "set-scrolling-region")
        ENTRY(CSI ";%dr", "set-scrolling-region-from-start")
        ENTRY(CSI "%dr", "set-scrolling-region-to-end")
        ENTRY(CSI "%d;r", "set-scrolling-region-to-end")
        ENTRY(CSI "%d;%dr", "set-scrolling-region")

        ENTRY(CSI "?%mr", "restore-mode")
        ENTRY(CSI "s", "save-cursor")
        ENTRY(CSI "?%ms", "save-mode")
        ENTRY(CSI "u", "restore-cursor")

        ENTRY(CSI "%mt", "window-manipulation")

        ENTRY(CSI "%d;%d;%d;%dw", "enable-filter-rectangle")
        ENTRY(CSI "%dx", "request-terminal-parameters")
        ENTRY(CSI "%d;%d'z", "enable-locator-reporting")
        ENTRY(CSI "%m'{", "select-locator-events")
        ENTRY(CSI "%d'|", "request-locator-position")

        COMMENT(/* Set text parameters, BEL-terminated versions. */)
        ENTRY(OSC ";%s" BEL, "set-icon-and-window-title") COMMENT(/* undocumented default */)
        ENTRY(OSC "0;%s" BEL, "set-icon-and-window-title")
        ENTRY(OSC "1;%s" BEL, "set-icon-title")
        ENTRY(OSC "2;%s" BEL, "set-window-title")
        ENTRY(OSC "3;%s" BEL, "set-xproperty")
        ENTRY(OSC "4;%s" BEL, "change-color-bel")
        ENTRY(OSC "6;%s" BEL, "set-current-file-uri")
        ENTRY(OSC "7;%s" BEL, "set-current-directory-uri")
        ENTRY(OSC "10;%s" BEL, "change-foreground-color-bel")
        ENTRY(OSC "11;%s" BEL, "change-background-color-bel")
        ENTRY(OSC "12;%s" BEL, "change-cursor-color-bel")
        ENTRY(OSC "13;%s" BEL, "change-mouse-cursor-foreground-color-bel")
        ENTRY(OSC "14;%s" BEL, "change-mouse-cursor-background-color-bel")
        ENTRY(OSC "15;%s" BEL, "change-tek-foreground-color-bel")
        ENTRY(OSC "16;%s" BEL, "change-tek-background-color-bel")
        ENTRY(OSC "17;%s" BEL, "change-highlight-background-color-bel")
        ENTRY(OSC "18;%s" BEL, "change-tek-cursor-color-bel")
        ENTRY(OSC "19;%s" BEL, "change-highlight-foreground-color-bel")
        ENTRY(OSC "46;%s" BEL, "change-logfile")
        ENTRY(OSC "50;#%d" BEL, "change-font-number")
        ENTRY(OSC "50;%s" BEL, "change-font-name")
        ENTRY(OSC "104" BEL, "reset-color")
        ENTRY(OSC "104;%m" BEL, "reset-color")
        ENTRY(OSC "110" BEL, "reset-foreground-color")
        ENTRY(OSC "111" BEL, "reset-background-color")
        ENTRY(OSC "112" BEL, "reset-cursor-color")
        ENTRY(OSC "113" BEL, "reset-mouse-cursor-foreground-color")
        ENTRY(OSC "114" BEL, "reset-mouse-cursor-background-color")
        ENTRY(OSC "115" BEL, "reset-tek-foreground-color")
        ENTRY(OSC "116" BEL, "reset-tek-background-color")
        ENTRY(OSC "117" BEL, "reset-highlight-background-color")
        ENTRY(OSC "118" BEL, "reset-tek-cursor-color")
        ENTRY(OSC "119" BEL, "reset-highlight-foreground-color")

        COMMENT(/* Set text parameters, ST-terminated versions. */)
        ENTRY(OSC ";%s" ST, "set-icon-and-window-title") COMMENT(/* undocumented default */)
        ENTRY(OSC "0;%s" ST, "set-icon-and-window-title")
        ENTRY(OSC "1;%s" ST, "set-icon-title")
        ENTRY(OSC "2;%s" ST, "set-window-title")
        ENTRY(OSC "3;%s" ST, "set-xproperty")
        ENTRY(OSC "4;%s" ST, "change-color-st")
        ENTRY(OSC "6;%s" ST, "set-current-file-uri")
        ENTRY(OSC "7;%s" ST, "set-current-directory-uri")
        ENTRY(OSC "10;%s" ST, "change-foreground-color-st")
        ENTRY(OSC "11;%s" ST, "change-background-color-st")
        ENTRY(OSC "12;%s" ST, "change-cursor-color-st")
        ENTRY(OSC "13;%s" ST, "change-mouse-cursor-foreground-color-st")
        ENTRY(OSC "14;%s" ST, "change-mouse-cursor-background-color-st")
        ENTRY(OSC "15;%s" ST, "change-tek-foreground-color-st")
        ENTRY(OSC "16;%s" ST, "change-tek-background-color-st")
        ENTRY(OSC "17;%s" ST, "change-highlight-background-color-st")
        ENTRY(OSC "18;%s" ST, "change-tek-cursor-color-st")
        ENTRY(OSC "19;%s" ST, "change-highlight-foreground-color-st")
        ENTRY(OSC "46;%s" ST, "change-logfile")
        ENTRY(OSC "50;#%d" ST, "change-font-number")
        ENTRY(OSC "50;%s" ST, "change-font-name")
        ENTRY(OSC "104" ST, "reset-color")
        ENTRY(OSC "104;%m" ST, "reset-color")
        ENTRY(OSC "110" ST, "reset-foreground-color")
        ENTRY(OSC "111" ST, "reset-background-color")
        ENTRY(OSC "112" ST, "reset-cursor-color")
        ENTRY(OSC "113" ST, "reset-mouse-cursor-foreground-color")
        ENTRY(OSC "114" ST, "reset-mouse-cursor-background-color")
        ENTRY(OSC "115" ST, "reset-tek-foreground-color")
        ENTRY(OSC "116" ST, "reset-tek-background-color")
        ENTRY(OSC "117" ST, "reset-highlight-background-color")
        ENTRY(OSC "118" ST, "reset-tek-cursor-color")
        ENTRY(OSC "119" ST, "reset-highlight-foreground-color")

        COMMENT(/* These may be bogus, I can't find docs for them anywhere (#104154). */)
        ENTRY(OSC "21;%s" BEL, "set-text-property-21")
        ENTRY(OSC "2L;%s" BEL, "set-text-property-2L")
        ENTRY(OSC "21;%s" ST, "set-text-property-21")
        ENTRY(OSC "2L;%s" ST, "set-text-property-2L")

        "\0";

#undef ENTRY
#undef COMMENT
